src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -238,19 +290,55 @@ class Wavefunction
         rng_.seed(s);
     }
 
+    void reorderWFN(unsigned qubitLoc, unsigned newPos) const
+    {
+        //get id of qubit located at newPos and qubitLoc - getting index from the element
+        auto itr1 = std::find(qubitmap_.begin(), qubitmap_.end(), newPos);
+        auto newPosId = std::distance(qubitmap_.begin(), itr1);
+        auto itr2 = std::find(qubitmap_.begin(), qubitmap_.end(), qubitLoc);
+        auto qid = std::distance(qubitmap_.begin(), itr2);
+        //swap qubits in wfn between qubitLoc and newPos
+        if (newPos != qubitLoc)
+        {
+            for (unsigned i = 0; i < wfn_.size(); i++)
+            {
+                unsigned int bit1 = (i >> qubitLoc) & 1;
+                unsigned int bit2 = (i >> newPos) & 1;
+                unsigned int x = (bit1 ^ bit2);
+                x = (x << qubitLoc) | (x << newPos);
+                unsigned new_i = i ^ x;
+                if (new_i > i)
+                {
+                    std::iter_swap(wfn_.begin() + i, wfn_.begin() + new_i);
+                }
+            }
+            //swap elemnts in qubitmap located at qid and newPosId
+            std::iter_swap(qubitmap_.begin() + qid, qubitmap_.begin() + newPosId);
+        }
+        
+    }
     /// generic application of a gate
     template <class Gate>
     void apply(Gate const& g)
     {
-        fused_.apply(wfn_, g.matrix(), qubit(g));
+        //check flush condition
+        std::vector<qubit_t> cs;
+        if (fused_.shouldFlush(wfn_, g.matrix(), cs, g.qubit())) {
```

Comment:
matrix() call creates the matrix, so now you are doing it twice -- maybe cache in a local instead?